### DISCUSSION TITLE
TechPrep : Javascript interview questions

### DISCUSSION THREAD

**Discussion by sharif-22** [anburocky3/anburocky3#5] [General]
1.  What are JavaScript Data Types?

    > - Number
    > - String
    > - Boolean
    > - Object
    > - Array
    > - Undefined
    > - Null
    > - Symbol

2.  What are undeclared and undefined variables?

    > Undeclared variables are those that do not exist in a program and are not declared. If the program tries to read the value of an undeclared variable, then a runtime error is encountered.

    > Undefined variables are those that are declared in the program but have not been given any value. If the program tries to read the value of an undefined variable, an undefined value is returned.

3.  Explain the difference between == and ===

    > "==" checks only for equality in value whereas "===" is a stricter equality test and returns false if either the value or the type of the two variables are different.

4.  What would be the result of 3+2+"7"?

    > Since 3 and 2 are integers, they will be added numerically. And since 7 is a string, its concatenation will be done. So the result would be 57.

5.  What are all the types of Pop up boxes available in JavaScript?

    > - Alert
    > - Confirm and
    > - Prompt

6.  What is the difference between an alert box and a confirmation box?

    > An alert box displays only one button which is the 0K button. But a Confirmation box displays two buttons namely 0K and cancel.

7.  What is the use of type of operator?

    > `Typeof` is an operator which is used to return a string description of the type of a variable.

8.  Is JavaScript case sensitive? Give an example?

    > Yes, JavaScript is case sensitive. For example, a function `parselnt` is not same as the function `Parseint`.

9.  What are callbacks in JavaScript, and how are they used?

    > Callbacks in JavaScript are functions that are passed as arguments to other functions and are executed at a later time or after an asynchronous operation has completed. They are used to ensure that certain code runs only after a particular task has finished executing.

    > For example, let's say you have a function called downloadFile that downloads a file from the internet. You can pass another function (a callback) to downloadFile as an argument. This callback function will be called once the file has finished downloading.

10. Explain the concept of function hoisting in JavaScript.

    > Function hoisting in JavaScript is a behavior where function declarations are moved to the top of their containing scope during the compilation phase. This means that you can call a function before it's actually declared in your code, and JavaScript will still know about it. However, it's important to note that only the function declarations, **not function expressions, are hoisted**. Understanding function hoisting helps developers organize their code more effectively and avoid unexpected behavior."

11. Explain the concept of scope in JavaScripts

    > Scope in JavaScript determines where variables and functions are accessible. Variables declared outside of functions have global scope, while those declared inside functions have local scope.

12. What is the difference between var, let, and const in JavaScript?

    > - `var` was traditionally used for variable declaration in JavaScript. It has function-level scope and can be re-declared and updated throughout the program.

    > - `let` and `const` were introduced in ES6. `let` allows block-level scope and enables variables to be re-assigned, but not re-declared within the same scope. `const` also has block-level scope, but it creates variables that cannot be re-assigned or re-declared after initialization.

    > - `let` and `const` provide better control over variable scoping and immutability compared to `var`, making code more predictable and maintainable.

13. What is the difference between setTimeout and setInterval functions in JavaScript?

    > - `setTimeout` is a function in JavaScript that allows you to execute a specified function or piece of code once after a specified delay in milliseconds.

    > - `setInterval`, on the other hand, is used to repeatedly execute a specified function or piece of code at a fixed interval (in milliseconds), starting after a specified delay.

    > - While `setTimeout` triggers the function or code block once after the specified delay, `setInterval` continues to trigger the function or code block at regular intervals until it is cleared using `clearInterval`.

    > - `setTimeout` is commonly used for delaying actions or executing code after a certain delay, while `setInterval` is used for executing code repeatedly at fixed intervals, such as in animations or periodic checks.

14. What are promises in JavaScript, and how do they work?

    > Promises in JavaScript are objects representing the eventual completion or failure of an asynchronous operation. They provide a cleaner way to handle asynchronous code compared to callback functions, making it easier to write and maintain asynchronous code.

    > A promise can be in one of three states: pending, fulfilled, or rejected. When a promise is pending, it means that the asynchronous operation is still ongoing. Once the operation is completed successfully, the promise transitions to the fulfilled state and returns a value. If an error occurs during the operation, the promise transitions to the rejected state and returns an error.

    > Promises have two main methods: `then()` and `catch()`. The `then()` method is used to handle the fulfillment of the promise, while the `catch()` method is used to handle any errors that occur during the operation. Additionally, promises can be chained together using `then()` to perform sequential asynchronous operations.

    > Overall, promises provide a more organized and readable way to work with asynchronous code, making it easier to manage complex asynchronous operations and handle errors effectively.

15. What is the purpose of the Array.isArray() method in JavaScript?

    > The `Array.isArray()` method in JavaScript is used to check if a value is an array. It returns `true` if the value is an array, and `false` otherwise. This is helpful when you need to make sure you're working with an array before performing array-specific operations.

16. What is the difference between map, filter, and reduce methods in JavaScript? Provide examples

    **`map()`:**

    - **Purpose:** Used for transforming each element of an array and creating a new array with the results.
    - **Example:**

      ```javascript
      const numbers = [1, 2, 3, 4, 5];

      const squaredNumbers = numbers.map((num) => num ** 2);

      console.log(squaredNumbers); // Output: [1, 4, 9, 16, 25]
      ```

    **`filter()`:**

    - **Purpose:** Used for creating a new array containing only elements that pass a certain condition.
    - **Example:**

      ```javascript
      const numbers = [1, 2, 3, 4, 5];

      const evenNumbers = numbers.filter((num) => num % 2 === 0);

      console.log(evenNumbers); // Output: [2, 4]
      ```

    **`reduce()`:**

    - **Purpose:** Used for accumulating values of an array into a single result. It takes a callback function that performs the accumulation.
    - **Example:**

      ```javascript
      const numbers = [1, 2, 3, 4, 5];

      const sum = numbers.reduce(
        (accumulator, currentValue) => accumulator + currentValue,
        0
      );

      console.log(sum); // Output: 15
      ```

    > In the `reduce` example, `accumulator` is the running total, `currentValue` is the current element being processed, and `0` is the initial value of the accumulator. The final result is the sum of all elements in the array.

    > In summary:

    > - `map` transforms each element and creates a new array.
    > - `filter` creates a new array with elements that pass a condition.
    > - `reduce` accumulates values into a single result.

17. What is the purpose of the window object in JavaScript, and how is it different from the document object?

    > The `window` object in JavaScript represents the browser window and provides access to browser-related functionalities like:
    >
    > - `setTimeout()` and `setInterval()` for timers.
    > - `window.open()` for opening new windows.
    > - `window.location` for accessing URL information.
    > - `localStorage` and `sessionStorage` for storage.
    > - `window.history` for navigation history.

    > On the other hand, the `document` object represents the HTML document loaded in the browser window and allows manipulation of the document's content and structure using methods like:
    >
    > - `getElementById()`, `getElementsByClassName()`, and `getElementsByTagName()` for accessing elements.
    > - `createElement()`, `appendChild()`, and `removeChild()` for modifying the document structure.
    > - `querySelector()` and `querySelectorAll()` for selecting elements using CSS selectors.

18. What is the difference between synchronous and asynchronous JavaScript code?

    > **Synchronous JavaScript Code:**
    >
    > - Executes line by line, in the order it appears in the code.
    > - Each operation waits for the previous one to finish before starting.
    > - Blocking: If an operation takes a long time to complete, it holds up the execution of subsequent code.

    > **Asynchronous JavaScript Code:**
    >
    > - Doesn't wait for each operation to finish before moving on to the next one.
    > - Uses callbacks, promises, or async/await to handle operations that may take time to complete.
    > - Non-blocking: Allows the program to continue executing other tasks while waiting for asynchronous operations to complete, improving performance and responsiveness.

19. What are higher-order functions in JavaScript, and how are they used?

    > **Higher-Order Functions:**
    >
    > - Higher-order functions in JavaScript are functions that can take other functions as arguments or return functions as results.
    >
    > - They enable more flexible and modular code by allowing functions to be treated as values and manipulated like any other data type.
    >
    > - Examples include `map()`, `filter()`, and `reduce()`, which take functions as arguments to operate on arrays, enhancing code readability and reusability.
    >
    > Example using `map()`:

    ```javascript

    const numbers = [1, 2, 3, 4];
    const squaredNumbers = numbers.map(function (num) {
    return num \* num;
    });
    // squaredNumbers: [1, 4, 9, 16]

    ```

    > In this example, `map()` is a higher-order function because it takes a function as an argument (`function(num) { return num * num;}`) to be applied to each element of the array.

20. What is the difference between the Array.prototype.map() and Array.prototype.forEach() methods in JavaScript?

    > **`Array.prototype.map()`:**
    >
    > - Creates a new array by applying a function to each element.
    > - Returns a transformed array.
    >
    > ```javascript
    > //Example:
    > const numbers = [1, 2, 3];
    > const doubledNumbers = numbers.map((num) => num \* 2);
    > // doubledNumbers: [2, 4, 6]
    > ```

    > **`Array.prototype.forEach()`:**
    >
    > - Executes a function for each element of an array.
    > - Doesn't create a new array.
    >
    > ```javascript
    > //Example:
    > const numbers = [1, 2, 3];
    > numbers.forEach((num) => console.log(num));
    > // Output:
    > // 1
    > // 2
    > // 3
    > ```

    > In short, `map()` transforms and produces a new array, while `forEach()` is used for executing a function on each array element without creating a new array.

**Comment by mrspecht** [comment]
Very useful. I would add a question about truthy/falsy values and  TDZ (Temporal Dead Zone)

**Reply by sharif-22** [reply]
Thank you for your feedback! That's a great suggestion.  Please feel free to include it in the comment section or create a new discussion and include all your questions and thoughts there its helpful for your  [Tech Prep app](https://github.com/anburocky3/tech-prep)

**Reply by mrspecht** [reply]
Will do for sure. Thanks

**Comment by mrspecht** [comment]
21. What is Temporal Dead Zone in JavaScript?

> In JavaScript, the temporal dead zone (TDZ) refers to the period between entering a scope (such as a function or block) and the actual declaration of a variable using let or const. During this period, accessing the variable will result in a ReferenceError. This occurs because variables declared with let and const are hoisted to the top of their scope, but they are not initialized until the actual declaration statement is encountered.
